app/models/manager_refresh/inventory_collection/builder/infra_manager.rb

```diff
@@ -0,0 +1,215 @@
+                    inventory_collection.store_updated_records(record)
+                  end
+
+                  inventory_object.id = record.id
+                end
+              end
+            end
+          )
+        end
+
+        def vms
+          super
+
+          add_custom_reconnect_block(
+            lambda do |inventory_collection, inventory_objects_index, attributes_index|
+              relation = inventory_collection.model_class.where(:ems_id => nil)
+
+              return if relation.count <= 0
+
+              inventory_objects_index.each_slice(100) do |batch|
+                relation.where(inventory_collection.manager_ref.first => batch.map(&:first)).each do |record|
+                  index = inventory_collection.object_index_with_keys(inventory_collection.manager_ref_to_cols, record)
+
+                  # We need to delete the record from the inventory_objects_index and attributes_index, otherwise it
+                  # would be sent for create.
+                  inventory_object = inventory_objects_index.delete(index)
+                  hash             = attributes_index.delete(index)
+
+                  record.assign_attributes(hash.except(:id, :type))
+                  if !inventory_collection.check_changed? || record.changed?
+                    record.save!
+                    inventory_collection.store_updated_records(record)
+                  end
+
+                  inventory_object.id = record.id
+                end
+              end
+            end
+          )
+        end
+
+        def host_storages
+          add_properties(
+            :manager_ref                  => %i(host storage),
+            :parent_inventory_collections => %i(hosts)
+          )
+        end
+
+        def host_switches
+          add_properties(
+            :manager_ref                  => %i(host switch),
+            :parent_inventory_collections => %i(hosts)
+          )
+        end
+
+        def switches
+          add_properties(
+            :manager_ref => %i(uid_ems)
+          )
+        end
+
+        def lans
+          add_properties(
+            :manager_ref => %i(uid_ems),
+          )
+        end
+
+        def snapshot_parent
+          snapshot_parent_save_block = lambda do |_ems, inventory_collection|
+            snapshot_collection = inventory_collection.dependency_attributes[:snapshots].try(:first)
+
+            snapshot_collection.each do |snapshot|
+              ActiveRecord::Base.transaction do
+                child = Snapshot.find(snapshot.id)
+                parent = Snapshot.find_by(:uid_ems => snapshot.parent_uid)
+                child.update_attribute(:parent_id, parent.try(:id))
+              end
+            end
+          end
+
+          add_properties(
+            :custom_save_block => snapshot_parent_save_block
+          )
+
+          # ???snapshot_patent??? in previous interface, probably mistake
+          # add_properties(:association => :snapshot_patent)
```

Comment:
rightn so the base class ```snapshot_parent``` should be ok, since there is a custom saving, I think this can be any name